### PR TITLE
sql: store partial index predicate in index descriptor

### PR DESCRIPTION
**Pull request by mgartner**
With this commit, serialized partial index predicates are now stored on
index descriptors. Predicates are dequalified so that database and table
names are not included in column references.

Release note: None

### CODE REVIEW

**Review by jordanlewis — CHANGES_REQUESTED** [review]
Nice, this looks good. I hope you don't mind the drop in reviews, but I'm just quite familiar with what's going on so it's pretty low cost for me.

I think that this patch will be broken in the presence of column renames. You need to change the code in `pkg/sql/rename_column.go` to iterate over the Predicates and not just the check constraints and computed column expressions. You should also add test cases for this, and also maybe some test cases for rename table as you allude to in one of your comments.

File: pkg/sql/partial_index.go
Comment (jordanlewis) [reply pkg/sql/partial_index.go:32]: Computed column expressions also fall into this category.

File: pkg/sql/partial_index.go
Comment (mgartner) [reply pkg/sql/partial_index.go:29]: Good catch. Accidentally left our subqueries. I'm not aware of any other things it does, though.

File: pkg/sql/create_table.go
Comment (mgartner) [reply pkg/sql/create_table.go:1481]: Good catch. I've made this change in the previous commit, here: https://github.com/cockroachdb/cockroach/pull/49423

File: pkg/sql/partial_index.go
Comment (mgartner) [reply pkg/sql/partial_index.go:32]: Great idea. I'm going to go ahead and merge this when ready, but I'll follow-up with a PR that reorganizes some of this stuff into a new package.

File: pkg/sql/partial_index.go
Comment (mgartner) [reply pkg/sql/partial_index.go:29]: I've made this change in the first commit, here: https://github.com/cockroachdb/cockroach/pull/49423

File: pkg/sql/partial_index.go
Comment (mgartner) [reply pkg/sql/partial_index.go]: I decided to try to tackle it now. @jordanlewis can you take a look at #49657 and let me know what you think? I was able to rip a lot of schema-expression related things out of sql and into a new package.

**Comment by mgartner** [comment]
Hold off on reviewing this until #49657 is sorted out. A rebase will be required if that gets merged.

File: pkg/sql/logictest/testdata/logic_test/partial_index
Comment (mgartner) [inline pkg/sql/logictest/testdata/logic_test/partial_index:1]: This is required for the `SHOW CREATE TABLE...` statements I added. Without it the build fails with `relation "system.zones" does not exist`. Why?

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49453#-:-M8mBFWuERHtf3KIktN8:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, @mgartner, and @RaduBerinde)

___
*[pkg/sql/logictest/testdata/logic_test/partial_index, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49453#-M7y3VDSXTRiBwq_igMj-r3-1:-M8mA6Hq63Tv8Bl2Amk-:blie5gy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c5aafdfb9410ae65fa72d8a4b2354fa1133444/pkg/sql/logictest/testdata/logic_test/partial_index#L1)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This is required for the `SHOW CREATE TABLE...` statements I added. Without it the build fails with `relation "system.zones" does not exist`. Why?
</blockquote></details>

Maybe @asubiotto knows. It looks like it's blacklisted in a lot of other tests so this is probably fine
___
*[pkg/sql/sqlbase/structured.proto, line 470 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49453#-M8mAIgeBEoDnYr1lH1H:-M8mAIgfDi4WMTVlT4gR:bxoep4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c5aafdfb9410ae65fa72d8a4b2354fa1133444/pkg/sql/sqlbase/structured.proto#L470)):*
> ```protobuf
>   optional geo.geoindex.Config geo_config = 22 [(gogoproto.nullable) = false];
> 
>   // If Predicate is not the empty string, describes a predicate expression for
> ```

[nit] It would be nice if it would be more clear that this is only set for partial indexes, and is actually the only way to tell if an index is partial. Maybe "Predicate, when it is not empty, indicates that this is a partial index and contains the expression for the partial index condition."
___
*[pkg/sql/sqlbase/structured.proto, line 471 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49453#-M8mB207ElUtSkBD5fDa:-M8mB207ElUtSkBD5fDb:b-vzjysq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c5aafdfb9410ae65fa72d8a4b2354fa1133444/pkg/sql/sqlbase/structured.proto#L471)):*
> ```protobuf
> 
>   // If Predicate is not the empty string, describes a predicate expression for
>   // a partial index.
> ```

[nit] maybe add a TODO referencing #49766. (once that is addressed, we'd want to update this comment to say how the predicate expression refers to table columns)
___
*[pkg/sql/sqlbase/structured.proto, line 472 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49453#-M8mAnCk3xH-HkcVYstp:-M8mAnCk3xH-HkcVYstq:bgmg4wg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c5aafdfb9410ae65fa72d8a4b2354fa1133444/pkg/sql/sqlbase/structured.proto#L472)):*
> ```protobuf
>   // If Predicate is not the empty string, describes a predicate expression for
>   // a partial index.
>   optional string predicate = 23 [(gogoproto.nullable) = false];
> ```

[nit] consider `partial_index_predicate` (your call, I realize it's a mouthful)

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49453)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @jordanlewis, and @RaduBerinde)

___
*[pkg/sql/sqlbase/structured.proto, line 470 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49453#-M8mAIgeBEoDnYr1lH1H:-M8mUH0ZA-RRUuslqK1K:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c5aafdfb9410ae65fa72d8a4b2354fa1133444/pkg/sql/sqlbase/structured.proto#L470)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] It would be nice if it would be more clear that this is only set for partial indexes, and is actually the only way to tell if an index is partial. Maybe "Predicate, when it is not empty, indicates that this is a partial index and contains the expression for the partial index condition."
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.proto, line 471 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49453#-M8mB207ElUtSkBD5fDa:-M8mUj8K4cCskDc2aj7O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c5aafdfb9410ae65fa72d8a4b2354fa1133444/pkg/sql/sqlbase/structured.proto#L471)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] maybe add a TODO referencing #49766. (once that is addressed, we'd want to update this comment to say how the predicate expression refers to table columns)
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.proto, line 472 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49453#-M8mAnCk3xH-HkcVYstp:-M8mUjus7YmgQh8RtZma:b-doeltg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c5aafdfb9410ae65fa72d8a4b2354fa1133444/pkg/sql/sqlbase/structured.proto#L472)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] consider `partial_index_predicate` (your call, I realize it's a mouthful)
</blockquote></details>

I felt that "index" was redundant given that this is within `IndexDescriptor`. Would `partial_predicate` be better than `predicate`?

**Comment by RaduBerinde** [comment]
___
*[pkg/sql/sqlbase/structured.proto, line 472 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49453#-M8mAnCk3xH-HkcVYstp:-M8mZxy59N4bLsFqeu5N:b-cknwra) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c5aafdfb9410ae65fa72d8a4b2354fa1133444/pkg/sql/sqlbase/structured.proto#L472)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I felt that "index" was redundant given that this is within `IndexDescriptor`. Would `partial_predicate` be better than `predicate`?
</blockquote></details>

Nah, that's more confusing (sounds like the predicate is partial). Better to keep it as is.

**Review by asubiotto — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49453)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis and @mgartner)

___
*[pkg/sql/logictest/testdata/logic_test/partial_index, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49453#-M7y3VDSXTRiBwq_igMj-r3-1:-M8o0xmY4g72umux4OJB:bk0p4q8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c5aafdfb9410ae65fa72d8a4b2354fa1133444/pkg/sql/logictest/testdata/logic_test/partial_index#L1)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Maybe @asubiotto knows. It looks like it's blacklisted in a lot of other tests so this is probably fine
</blockquote></details>

Because system.zones only exists on the system tenant: https://github.com/cockroachdb/cockroach/issues/49445. https://github.com/cockroachdb/cockroach/pull/49784 should make it so that we can remove this blacklist for a lot of tests that depend on accessing zone configs. Blacklisting is fine.

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49453)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis)

___
*[pkg/sql/logictest/testdata/logic_test/partial_index, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49453#-M7y3VDSXTRiBwq_igMj-r3-1:-M8ptXUxAk7ewvT1SS6A:b-wsvasg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c5aafdfb9410ae65fa72d8a4b2354fa1133444/pkg/sql/logictest/testdata/logic_test/partial_index#L1)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Because system.zones only exists on the system tenant: https://github.com/cockroachdb/cockroach/issues/49445. https://github.com/cockroachdb/cockroach/pull/49784 should make it so that we can remove this blacklist for a lot of tests that depend on accessing zone configs. Blacklisting is fine.
</blockquote></details>

Thanks for the explanation!
___
*[pkg/sql/sqlbase/structured.proto, line 472 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49453#-M8mAnCk3xH-HkcVYstp:-M8ptacb27LgO8y_uc08:bp2jobb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c5aafdfb9410ae65fa72d8a4b2354fa1133444/pkg/sql/sqlbase/structured.proto#L472)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Nah, that's more confusing (sounds like the predicate is partial). Better to keep it as is.
</blockquote></details>

Ok. I'll keep an eye on this, and if it is confusing us again before the release, I can rename it.

**Pull request merged**
